<!-- ---------------------------------------------------- --> 
  <div class="col-sm-6 col-md-6 col-lg-6">
	<!-- Pandas Tutorial -->  
	  <center>
	    <br>
        <h1>Part 2</h1>
        <br>
        <br>
	  </center>	
    <!-- Pandas Tutorial -->	
  </div>
<!-- ---------------------------------------------------- -->

**Get Data** - Our data set will consist of an Excel file containing customer counts per date. We will learn how to read in the excel file for processing.  
**Prepare Data** - The data is an irregular time series having duplicate dates. We will be challenged in compressing the data and coming up with next years forecasted customer count.  
**Analyze Data** - We use graphs to visualize trends and spot outliers. Some built in computational tools will be used to calculate next years forecasted customer count.  
**Present Data** - The results will be plotted.  

In [2]:
# Import libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy.random as np
import sys
import matplotlib

%matplotlib inline

> We will be creating our own test data for analysis.

In [38]:
# set seed
np.seed(111)

# Function to generate test data
def CreateDataSet(Number=1):
    
    Output = []
    
    for i in range(Number):
        
        # Create a weekly (mondays) date range
        rng = pd.date_range(start='1/1/2017', end='11/20/2017', freq='W-MON')
        
        # Create random data
        data = np.randint(low=25,high=1000,size=len(rng))
        
        # Status pool 
        status = [1,2,3]
        
        # Make a random list of statuses
        random_status = [status[np.randint(low=0,high=len(status))] for i in range(len(rng))]
        
        # State pool
        countries = ['BEL','DEU','AUS','NGA','PHL','ECU', 'BRA','JPN']
        
        # Make a random list of states 
        random_countries = [countries[np.randint(low=0,high=len(countries))] for i in range(len(rng))]
    
        Output.extend(zip(random_countries, random_status, data, rng))
        
    return Output

Now that we have a function to generate our test data, lets create some data and stick it into a dataframe.

In [39]:
dataset = CreateDataSet(4)
df = pd.DataFrame(data=dataset, columns=['Countries','Product','CustomerCount','StatusDate'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188 entries, 0 to 187
Data columns (total 4 columns):
Countries        188 non-null object
Product          188 non-null int64
CustomerCount    188 non-null int64
StatusDate       188 non-null datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 5.9+ KB


In [40]:
df

Countries  Product  CustomerCount StatusDate
0         NGA        1            877 2017-01-02
1         PHL        2            901 2017-01-09
2         DEU        1            749 2017-01-16
3         DEU        2            111 2017-01-23
4         ECU        3            300 2017-01-30
5         BEL        2            706 2017-02-06
6         ECU        1            347 2017-02-13
7         ECU        2            143 2017-02-20
8         PHL        3            833 2017-02-27
9         BEL        1            992 2017-03-06
10        AUS        3            978 2017-03-13
11        ECU        3             32 2017-03-20
12        ECU        2            293 2017-03-27
13        BRA        3            291 2017-04-03
14        DEU        2            474 2017-04-10
15        BEL        1            905 2017-04-17
16        AUS        1            753 2017-04-24
17        JPN        2            309 2017-05-01
18        NGA        2            295 2017-05-08
19        ECU        1            443 2017-05-15
20        AUS        1             46 2017-05-22
21        BRA        2            591 2017-05-29
22        PHL        2            737 2017-06-05
23        JPN        2             62 2017-06-12
24        JPN        2            997 2017-06-19
25        NGA        1            767 2017-06-26
26        NGA        2            723 2017-07-03
27        BEL        1            293 2017-07-10
28        NGA        1            710 2017-07-17
29        BRA        3            990 2017-07-24
..        ...      ...            ...        ...
158       NGA        2            671 2017-05-01
159       BRA        1            452 2017-05-08
160       PHL        2            604 2017-05-15
161       AUS        1            340 2017-05-22
162       ECU        2            377 2017-05-29
163       BEL        3             75 2017-06-05
164       AUS        3            377 2017-06-12
165       ECU        2            228 2017-06-19
166       NGA        1            928 2017-06-26
167       PHL        3            166 2017-07-03
168       ECU        1            666 2017-07-10
169       ECU        3            279 2017-07-17
170       JPN        2             73 2017-07-24
171       ECU        2            338 2017-07-31
172       PHL        2            479 2017-08-07
173       PHL        2            242 2017-08-14
174       BEL        1            128 2017-08-21
175       DEU        1            161 2017-08-28
176       AUS        3             63 2017-09-04
177       NGA        1            994 2017-09-11
178       ECU        3            744 2017-09-18
179       ECU        3            648 2017-09-25
180       BEL        2            638 2017-10-02
181       DEU        1            267 2017-10-09
182       DEU        3            991 2017-10-16
183       DEU        2            623 2017-10-23
184       BRA        1            978 2017-10-30
185       AUS        3            492 2017-11-06
186       NGA        3            256 2017-11-13
187       BRA        1            635 2017-11-20

[188 rows x 4 columns]

In [41]:
df.head() #GETTING First 5 rows

Countries  Product  CustomerCount StatusDate
0       NGA        1            877 2017-01-02
1       PHL        2            901 2017-01-09
2       DEU        1            749 2017-01-16
3       DEU        2            111 2017-01-23
4       ECU        3            300 2017-01-30

We are now going to save this dataframe into an Excel file, to then bring it back to a dataframe. We simply do this to show you how to read and write to Excel files.  

We do not write the index values of the dataframe to the Excel file, since they are not meant to be part of our initial test data set.

In [42]:
# Save results to excel
df.to_excel('Lesson3.xlsx', index=False)
print('Done')

Done


# Grab Data from Excel  

We will be using the ***read_excel*** function to read in data from an Excel file. The function allows you to read in specfic tabs by name or location.

In [28]:
pd.read_excel?

**Note: The location on the Excel file will be in the same folder as the notebook, unless specified otherwise.**

In [43]:
# Location of file
Location = r'Lesson3.xlsx'
# Parse a specific sheet
df = pd.read_excel(Location, 0, index_col='StatusDate')
df.dtypes

Countries        object
Product           int64
CustomerCount     int64
dtype: object

In [44]:
df

Countries  Product  CustomerCount
StatusDate                                  
2017-01-02       NGA        1            877
2017-01-09       PHL        2            901
2017-01-16       DEU        1            749
2017-01-23       DEU        2            111
2017-01-30       ECU        3            300
2017-02-06       BEL        2            706
2017-02-13       ECU        1            347
2017-02-20       ECU        2            143
2017-02-27       PHL        3            833
2017-03-06       BEL        1            992
2017-03-13       AUS        3            978
2017-03-20       ECU        3             32
2017-03-27       ECU        2            293
2017-04-03       BRA        3            291
2017-04-10       DEU        2            474
2017-04-17       BEL        1            905
2017-04-24       AUS        1            753
2017-05-01       JPN        2            309
2017-05-08       NGA        2            295
2017-05-15       ECU        1            443
2017-05-22       AUS        1             46
2017-05-29       BRA        2            591
2017-06-05       PHL        2            737
2017-06-12       JPN        2             62
2017-06-19       JPN        2            997
2017-06-26       NGA        1            767
2017-07-03       NGA        2            723
2017-07-10       BEL        1            293
2017-07-17       NGA        1            710
2017-07-24       BRA        3            990
...              ...      ...            ...
2017-05-01       NGA        2            671
2017-05-08       BRA        1            452
2017-05-15       PHL        2            604
2017-05-22       AUS        1            340
2017-05-29       ECU        2            377
2017-06-05       BEL        3             75
2017-06-12       AUS        3            377
2017-06-19       ECU        2            228
2017-06-26       NGA        1            928
2017-07-03       PHL        3            166
2017-07-10       ECU        1            666
2017-07-17       ECU        3            279
2017-07-24       JPN        2             73
2017-07-31       ECU        2            338
2017-08-07       PHL        2            479
2017-08-14       PHL        2            242
2017-08-21       BEL        1            128
2017-08-28       DEU        1            161
2017-09-04       AUS        3             63
2017-09-11       NGA        1            994
2017-09-18       ECU        3            744
2017-09-25       ECU        3            648
2017-10-02       BEL        2            638
2017-10-09       DEU        1            267
2017-10-16       DEU        3            991
2017-10-23       DEU        2            623
2017-10-30       BRA        1            978
2017-11-06       AUS        3            492
2017-11-13       NGA        3            256
2017-11-20       BRA        1            635

[188 rows x 3 columns]

In [45]:
df.index

DatetimeIndex(['2017-01-02', '2017-01-09', '2017-01-16', '2017-01-23',
               '2017-01-30', '2017-02-06', '2017-02-13', '2017-02-20',
               '2017-02-27', '2017-03-06',
               ...
               '2017-09-18', '2017-09-25', '2017-10-02', '2017-10-09',
               '2017-10-16', '2017-10-23', '2017-10-30', '2017-11-06',
               '2017-11-13', '2017-11-20'],
              dtype='datetime64[ns]', name=u'StatusDate', length=188, freq=None)

In [46]:
df.head()

Countries  Product  CustomerCount
StatusDate                                  
2017-01-02       NGA        1            877
2017-01-09       PHL        2            901
2017-01-16       DEU        1            749
2017-01-23       DEU        2            111
2017-01-30       ECU        3            300

# Prepare Data  

We first need to prepare the data for analysis
1. Make sure the countries column is all in upper case  
2. Only select records where the account status is equal to "1"  
3. Merge (NJ and NY) to NY in the state column  
4. Remove any outliers (any odd results in the data set)


Lets take a quick look on how some of the *State* values are upper case and some are lower case

In [47]:
df['Countries'].unique()

array([u'NGA', u'PHL', u'DEU', u'ECU', u'BEL', u'AUS', u'BRA', u'JPN'], dtype=object)

To convert all the State values to upper case we will use the ***upper()*** function and the dataframe's ***apply*** attribute. The ***lambda*** function simply will apply the upper function to each value in the *State* column.

In [48]:
# Clean State Column, convert to upper case
df['Countries'] = df.Countries.apply(lambda x: x.upper())

In [50]:
df['Countries'].unique()

array([u'NGA', u'PHL', u'DEU', u'ECU', u'BEL', u'AUS', u'BRA', u'JPN'], dtype=object)

In [52]:
# Only grab where Status == 1
mask = df['Product'] == 1
df = df[mask]

# Lets Group European Countries to EU

To turn the ***DEU***, ***BEL*** states to ***EU*** we simply...  

* ***[df.Countries == 'DEU']*** & ***[df.Countries == 'BEL']*** - Find all records in the *Countries* column where they are equal to *EU*.  
* ***df.Countries[df.Countries == 'DEU'] = 'EU'*** 
* ***df.Countries[df.Countries == 'BEL'] = 'EU'*** - For all records in the *Coutries* column where they are equal to *DEU* & *BEL*, replace them with *EU*.

In [55]:
# Convert DEU TO EU
mask = df.Countries == 'DEU'
df['Countries'][mask]

/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()
//anaconda/lib/python2.7/site-packages/pandas/core/generic.py:5088: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
//anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2885: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


Now we can see we have a much cleaner data set to work with.

In [ ]:
df['State'].unique()

At this point we may want to graph the data to check for any outliers or inconsistencies in the data. We will be using the ***plot()*** attribute of the dataframe.  

As you can see from the graph below it is not very conclusive and is probably a sign that we need to perform some more data preparation.

In [ ]:
df['CustomerCount'].plot(figsize=(15,5));

If we take a look at the data, we begin to realize that there are multiple values for the same State, StatusDate, and Status combination. It is possible that this means the data you are working with is dirty/bad/inaccurate, but we will assume otherwise. We can assume this data set is a subset of a bigger data set and if we simply add the values in the ***CustomerCount*** column per State, StatusDate, and Status we will get the ***Total Customer Count*** per day.  

In [ ]:
sortdf = df[df['State']=='NY'].sort_index(axis=0)
sortdf.head(10)

Our task is now to create a new dataframe that compresses the data so we have daily customer counts per State and StatusDate. We can ignore the Status column since all the values in this column are of value *1*. To accomplish this we will use the dataframe's functions ***groupby*** and ***sum()***.  

Note that we had to use **reset_index** . If we did not, we would not have been able to group by both the State and the StatusDate since the groupby function expects only columns as inputs. The **reset_index** function will bring the index ***StatusDate*** back to a column in the dataframe. 

In [ ]:
# Group by State and StatusDate
Daily = df.reset_index().groupby(['State','StatusDate']).sum()
Daily.head()

The ***State*** and ***StatusDate*** columns are automatically placed in the index of the ***Daily*** dataframe. You can think of the ***index*** as the primary key of a database table but without the constraint of having unique values. Columns in the index as you will see allow us to easily select, plot, and perform calculations on the data.  

Below we delete the ***Status*** column since it is all equal to one and no longer necessary.

In [ ]:
del Daily['Status']
Daily.head()

In [ ]:
# What is the index of the dataframe
Daily.index

In [ ]:
# Select the State index
Daily.index.levels[0]

In [ ]:
# Select the StatusDate index
Daily.index.levels[1]

Lets now plot the data per State.  

As you can see by breaking the graph up by the ***State*** column we have a much clearer picture on how the data looks like. Can you spot any outliers?

In [ ]:
Daily.loc['FL'].plot()
Daily.loc['GA'].plot()
Daily.loc['NY'].plot()
Daily.loc['TX'].plot();

We can also just plot the data on a specific date, like ***2012***. We can now clearly see that the data for these states is all over the place. since the data consist of weekly customer counts, the variability of the data seems suspect. For this tutorial we will assume bad data and proceed. 

In [ ]:
Daily.loc['FL']['2012':].plot()
Daily.loc['GA']['2012':].plot()
Daily.loc['NY']['2012':].plot()
Daily.loc['TX']['2012':].plot();

We will assume that per month the customer count should remain relatively steady. Any data outside a specific range in that month will be removed from the data set. The final result should have smooth graphs with no spikes.  

***StateYearMonth*** - Here we group by State, Year of StatusDate, and Month of StatusDate.  
***Daily['Outlier']*** - A boolean (True or False) value letting us know if the value in the CustomerCount column is ouside the acceptable range.  

We will be using the attribute ***transform*** instead of ***apply***. The reason is that transform will keep the shape(# of rows and columns) of the dataframe the same and apply will not. By looking at the previous graphs, we can realize they are not resembling a gaussian distribution, this means we cannot use summary statistics like the mean and stDev. We use percentiles instead. Note that we run the risk of eliminating good data.

In [ ]:
# Calculate Outliers
StateYearMonth = Daily.groupby([Daily.index.get_level_values(0), Daily.index.get_level_values(1).year, Daily.index.get_level_values(1).month])
Daily['Lower'] = StateYearMonth['CustomerCount'].transform( lambda x: x.quantile(q=.25) - (1.5*x.quantile(q=.75)-x.quantile(q=.25)) )
Daily['Upper'] = StateYearMonth['CustomerCount'].transform( lambda x: x.quantile(q=.75) + (1.5*x.quantile(q=.75)-x.quantile(q=.25)) )
Daily['Outlier'] = (Daily['CustomerCount'] < Daily['Lower']) | (Daily['CustomerCount'] > Daily['Upper']) 

# Remove Outliers
Daily = Daily[Daily['Outlier'] == False]

The dataframe named ***Daily*** will hold customer counts that have been aggregated per day. The original data (df) has multiple records per day.  We are left with a data set that is indexed by both the state and the StatusDate. The Outlier column should be equal to ***False*** signifying that the record is not an outlier.

In [ ]:
Daily.head()

We create a separate dataframe named ***ALL*** which groups the Daily dataframe by StatusDate. We are essentially getting rid of the ***State*** column. The ***Max*** column represents the maximum customer count per month. The ***Max*** column is used to smooth out the graph.

In [ ]:
# Combine all markets

# Get the max customer count by Date
ALL = pd.DataFrame(Daily['CustomerCount'].groupby(Daily.index.get_level_values(1)).sum())
ALL.columns = ['CustomerCount'] # rename column

# Group by Year and Month
YearMonth = ALL.groupby([lambda x: x.year, lambda x: x.month])

# What is the max customer count per Year and Month
ALL['Max'] = YearMonth['CustomerCount'].transform(lambda x: x.max())
ALL.head()

As you can see from the ***ALL*** dataframe above, in the month of January 2009, the maximum customer count was 901. If we had used ***apply***, we would have got a dataframe with (Year and Month) as the index and just the *Max* column with the value of 901. 

----------------------------------  
There is also an interest to gauge if the current customer counts were reaching certain goals the company had established. The task here is to visually show if the current customer counts are meeting the goals listed below. We will call the goals ***BHAG*** (Big Hairy Annual Goal).  

* 12/31/2011 - 1,000 customers  
* 12/31/2012 - 2,000 customers  
* 12/31/2013 - 3,000 customers  

We will be using the **date_range** function to create our dates.  

***Definition:*** date_range(start=None, end=None, periods=None, freq='D', tz=None, normalize=False, name=None, closed=None)  
***Docstring:*** Return a fixed frequency datetime index, with day (calendar) as the default frequency  

By choosing the frequency to be ***A*** or annual we will be able to get the three target dates from above.

In [ ]:
pd.date_range?

In [ ]:
# Create the BHAG dataframe
data = [1000,2000,3000]
idx = pd.date_range(start='12/31/2011', end='12/31/2013', freq='A')
BHAG = pd.DataFrame(data, index=idx, columns=['BHAG'])
BHAG

Combining dataframes as we have learned in previous lesson is made simple using the ***concat*** function. Remember when we choose ***axis = 0*** we are appending row wise.

In [ ]:
# Combine the BHAG and the ALL data set 
combined = pd.concat([ALL,BHAG], axis=0)
combined = combined.sort_index(axis=0)
combined.tail()

In [ ]:
fig, axes = plt.subplots(figsize=(12, 7))

combined['BHAG'].fillna(method='pad').plot(color='green', label='BHAG')
combined['Max'].plot(color='blue', label='All Markets')
plt.legend(loc='best');

There was also a need to forecast next year's customer count and we can do this in a couple of simple steps. We will first group the ***combined*** dataframe by ***Year*** and place the maximum customer count for that year. This will give us one row per Year.   

In [ ]:
# Group by Year and then get the max value per year
Year = combined.groupby(lambda x: x.year).max()
Year

In [ ]:
# Add a column representing the percent change per year
Year['YR_PCT_Change'] = Year['Max'].pct_change(periods=1)
Year

To get next year's end customer count we will assume our current growth rate remains constant. We then will increase this years customer count by that amount and that will be our forecast for next year. 

In [ ]:
(1 + Year.ix[2012,'YR_PCT_Change']) * Year.loc[2012,'Max']

# Present Data  

Create individual Graphs per State.  

In [ ]:
# First Graph
ALL['Max'].plot(figsize=(10, 5));plt.title('ALL Markets')

# Last four Graphs
fig, axes = plt.subplots(nrows=2, ncols=2, figsize=(20, 10))
fig.subplots_adjust(hspace=1.0) ## Create space between plots

Daily.loc['FL']['CustomerCount']['2012':].fillna(method='pad').plot(ax=axes[0,0])
Daily.loc['GA']['CustomerCount']['2012':].fillna(method='pad').plot(ax=axes[0,1]) 
Daily.loc['TX']['CustomerCount']['2012':].fillna(method='pad').plot(ax=axes[1,0]) 
Daily.loc['NY']['CustomerCount']['2012':].fillna(method='pad').plot(ax=axes[1,1]) 

# Add titles
axes[0,0].set_title('Florida')
axes[0,1].set_title('Georgia')
axes[1,0].set_title('Texas')
axes[1,1].set_title('North East');

<p class="text-muted">This tutorial was created by <a href="http://www.hedaro.com" target="_blank"><strong>HEDARO</strong></a></p>